optimum/bettertransformer/models/encoder_models.py

```
@@ -545,6 +545,136 @@ def forward(self, hidden_states, attention_mask, position_bias=None, *_, **__):
+            "norm1_eps": "self_attn_layer_norm.eps",
+            "norm1_weight": "self_attn_layer_norm.weight",
+            "norm1_bias": "self_attn_layer_norm.bias",
+            "norm2_eps": "final_layer_norm.eps",
+            "norm2_weight": "final_layer_norm.weight",
+            "norm2_bias": "final_layer_norm.bias",
+        }
+
+        self.validate_bettertransformer()
+
+    def forward(self, hidden_states, attention_mask, *_, **__):
+        super().forward_checker()
+
+        if not hasattr(hidden_states, "original_shape"):
+            original_shape = hidden_states.shape
+        else:
+            original_shape = hidden_states.original_shape
+
+        if hidden_states.is_nested:
+            attention_mask = None
+
+        if attention_mask is not None:
+            # attention mask comes in with values 0 and -inf. we convert to torch.nn.TransformerEncoder style bool mask
+            # 0->false->keep this token -inf->true->mask this token
+            attention_mask = attention_mask.squeeze(1)[:, 0]
+            attention_mask = attention_mask.bool()
+            attention_mask = torch.reshape(attention_mask, (attention_mask.shape[0], attention_mask.shape[-1]))
+            hidden_states = torch._nested_tensor_from_mask(hidden_states, ~attention_mask)
+            attention_mask = None
+
+        hidden_states = torch._transformer_encoder_layer_fwd(
+            hidden_states,
+            self.embed_dim,
+            self.num_heads,
+            self.in_proj_weight,
+            self.in_proj_bias,
+            self.out_proj_weight,
+            self.out_proj_bias,
+            self.norm_first,
+            self.norm1_eps,
+            self.norm1_weight,
+            self.norm1_bias,
+            self.norm2_eps,
+            self.norm2_weight,
+            self.norm2_bias,
+            self.linear1_weight,
+            self.linear1_bias,
+            self.linear2_weight,
+            self.linear2_bias,
+            attention_mask,
+        )
```

@awinml you forgot `self.use_gelu`, I guess that's why your test is failing.

---

@IlyasMoutawwakil I ran the tests after adding `self.use_gelu`. The above test (https://github.com/huggingface/optimum/pull/1065#issuecomment-1565458389) still fails.

---

the order in `torch._transformer_encoder_layer_fwd` is very important because there are no default values (almost).
`use_gelu` tells the cpp kernel to whether use gelu or relu (true, false). any other activation function should raise an error here:
https://github.com/huggingface/optimum/blob/eb05de1f4abe34ebc23c638780934ebd2fe3bb02/optimum/bettertransformer/models/base.py#LL116-L119
but apparently it doesn't and that's why your test doesn't pass (it's a sanity check).

---

@IlyasMoutawwakil Makes sense! I updated the order to match the link. The test still seems to fail. Can you have a look?

---

@younesbelkada Any suggestions for fixing the failing test? https://github.com/huggingface/optimum/pull/1065#issuecomment-1565458389